src/main/java/com/google/cloud/spanner/pgadapter/wireoutput/DataRowResponse.java

```diff
@@ -0,0 +1,100 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.cloud.spanner.pgadapter.wireoutput;
+
+import com.google.cloud.spanner.pgadapter.ConnectionHandler.QueryMode;
+import com.google.cloud.spanner.pgadapter.ProxyServer.DataFormat;
+import com.google.cloud.spanner.pgadapter.metadata.OptionsMetadata;
+import com.google.cloud.spanner.pgadapter.statements.IntermediateStatement;
+import com.google.cloud.spanner.pgadapter.utils.Converter;
+import java.io.DataOutputStream;
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Sends to the client specific row contents.
+ */
+public class DataRowResponse extends WireOutput {
+
+  private static final int HEADER_LENGTH = 4;
+  private static final int COLUMN_NUMBER_LENGTH = 2;
+  private static final int COLUMN_SIZE_LENGTH = 4;
+
+
+  private final IntermediateStatement statement;
+  private final QueryMode mode;
+  private final ResultSet resultSet;
+  private final ResultSetMetaData metadata;
+  private final List<byte[]> columns;
+
+  public DataRowResponse(DataOutputStream output,
+      IntermediateStatement statement,
+      OptionsMetadata options,
+      QueryMode mode) throws Exception {
+    super(output, 0);
+    this.statement = statement;
+    this.mode = mode;
+    this.resultSet = statement.getStatementResult();
+    this.metadata = resultSet.getMetaData();
+    this.columns = new ArrayList<>(this.metadata.getColumnCount());
+    this.length = HEADER_LENGTH + COLUMN_NUMBER_LENGTH;
+    for (int column_index = 1; /* column indices start at 1 */
+        column_index <= this.metadata.getColumnCount();
+        column_index++) {
+      length += COLUMN_SIZE_LENGTH;
+      if (resultSet.getObject(column_index) == null) {
+        columns.add(null);
+      } else {
+        DataFormat format = DataFormat.getDataFormat(column_index, statement, this.mode, options);
+        byte[] column
+            = new Converter().parseData(this.resultSet, this.metadata, column_index, format);
```

Comment:
this looks like the tight loop for sending data back from the proxy?  Do we need to pay attention to object creation counts here, as spooling a large result set would exercise this code a lot.

Reply:
There is a maximum number of columns per response (PG accept 1600, but Spanner is 1024, so it should exceed that), so I don't think it should blow up enough to cause too much damage to memory, considering we would also be taking each of these fields into memory - I can add a todo to profile it later :)